# UCL AI Society Machine Learning Tutorials
### Session 01. Introduction to Numpy, Pandas, Matplotlib

### Contents
1. Numpy
2. Pandas
3. Matplotlib
4. EDA(Exploratory Data Analysis)

### Aim
At the end of this session, you will be able to:
- Understand the basics of numpy.
- Understand the basics of pandas.
- Understand the basics of matplotlib.
- Perform an Exploratory Data Analysis (EDA).


## 2. Pandas
Pandas is another essential open-source library in Python, and today it is widely used by data scientists and ML engineers. It is built by Wes McKinney and is based on NumPy. The name 'Pandas' originates from the term "Panel Data", an econometrics term for datasets that include observations over multiple time periods for the same object.

### 2.1 Basics of Pandas

In [ ]:
# run this shell if you haven't installed pandas library
! pip install pandas

In [ ]:
import pandas as pd
import numpy as np

In [ ]:
print(pd.__version__)

The main data structures of pandas are **Series** and **DataFrame**, where data are stored and manipulated. A `Series` can simply be understood as a column and a `DataFrame` as a table that has many Series.

In [4]:
a = pd.Series([1, 2, 3, np.nan, 5, 6])

In [5]:
a

0    1.0
1    2.0
2    3.0
3    NaN
4    5.0
5    6.0
dtype: float64

**Let's see how they are different!**

In [6]:
# Creating a Series using pandas.Series()
# This is just one of many ways to initialise a pandas series
module_score_dic = {'Database': 90, 'Security': 70, 'Math': 100, 'Machine Learning': 80}
module_score = pd.Series(module_score_dic)
print("Module_score: \n", module_score, '\n')
print("type: ", type(module_score), '\n')

# Creating a DataFrame using pandas DataFrame()
dataframe = pd.DataFrame(module_score, columns=['score'])
# dataframe = pd.DataFrame(module_score, index=[x for x in module_score.keys()], columns=['score'])
print("dataframe: \n", dataframe, '\n')
print("type: ", type(dataframe))

Module_score: 
 Database             90
Security             70
Math                100
Machine Learning     80
dtype: int64 

type:  <class 'pandas.core.series.Series'> 

dataframe: 
                   score
Database             90
Security             70
Math                100
Machine Learning     80 

type:  <class 'pandas.core.frame.DataFrame'>


Series can also be a Dataframe that has only one attribute.  
**Now let's make a Dataframe that has multiple attributes**

In [7]:
solar_data = {
    'Name' : ["Mercury", "Venus", "Earth", "Mars", "Jupiter", "Saturn", "Uranus", "Neptune"],
    'Satellite' : [0, 0, 1, 2, 79, 60, 27, 14],
    'AU' : [0.4, 0.7, 1, 1.5, 5.2, 9.5, 19.2, 30.1],
    'Diameter (in 1Kkm)' : [4.9, 12.1, 12.7, 6.8, 139.8, 116.5, 50.7, 49.2]
}

In [8]:
solar_system = pd.DataFrame(solar_data, index = [i for i in range(1, 9)])
solar_system

Name  Satellite    AU  Diameter (in 1Kkm)
1  Mercury          0   0.4                 4.9
2    Venus          0   0.7                12.1
3    Earth          1   1.0                12.7
4     Mars          2   1.5                 6.8
5  Jupiter         79   5.2               139.8
6   Saturn         60   9.5               116.5
7   Uranus         27  19.2                50.7
8  Neptune         14  30.1                49.2

In [9]:
solar_system.dtypes # checks data type

Name                   object
Satellite               int64
AU                    float64
Diameter (in 1Kkm)    float64
dtype: object

We can select what to read from the DataFrame by using methods that df.DataFrame has.

- `head()` : Returns the first n data
- `tail()` : Returns the last n data
- `index` : Returns the index
- `columns` : Returns the column
- `loc` : Returns the information of that row
- `values` : Returns only the values without index and column names
- `describe()` : Outputs satatistical summary a DataFrame
- `sort_values(self, by, axis = 0, ascending = True, inplace = False)` : Sort the DataFrame
- `drop()` : Drops selected row

In [10]:
df = solar_system

In [12]:
df.head() # the default value in the brackets is 5

Name  Satellite   AU  Diameter (in 1Kkm)
1  Mercury          0  0.4                 4.9
2    Venus          0  0.7                12.1
3    Earth          1  1.0                12.7
4     Mars          2  1.5                 6.8
5  Jupiter         79  5.2               139.8

In [13]:
df.tail(2)

Name  Satellite    AU  Diameter (in 1Kkm)
7   Uranus         27  19.2                50.7
8  Neptune         14  30.1                49.2

In [14]:
df.index

Int64Index([1, 2, 3, 4, 5, 6, 7, 8], dtype='int64')

In [15]:
df.index[0]

1

In [16]:
df.columns

Index(['Name', 'Satellite', 'AU', 'Diameter (in 1Kkm)'], dtype='object')

In [17]:
df.loc[1]

Name                  Mercury
Satellite                   0
AU                        0.4
Diameter (in 1Kkm)        4.9
Name: 1, dtype: object

In [18]:
df.iloc[0]

Name                  Mercury
Satellite                   0
AU                        0.4
Diameter (in 1Kkm)        4.9
Name: 1, dtype: object

In [ ]:
# TODO: run this cell and observe it gives you an error. Do you see why? 
# Try playing around with different numbers and find out why before you google it!
# Google search keyword: df.loc vs df.iloc

df.loc[0]

In [19]:
df.values

array([['Mercury', 0, 0.4, 4.9],
       ['Venus', 0, 0.7, 12.1],
       ['Earth', 1, 1.0, 12.7],
       ['Mars', 2, 1.5, 6.8],
       ['Jupiter', 79, 5.2, 139.8],
       ['Saturn', 60, 9.5, 116.5],
       ['Uranus', 27, 19.2, 50.7],
       ['Neptune', 14, 30.1, 49.2]], dtype=object)

In [20]:
df.describe()

Satellite         AU  Diameter (in 1Kkm)
count   8.000000   8.000000             8.00000
mean   22.875000   8.450000            49.08750
std    30.670775  10.853702            52.38417
min     0.000000   0.400000             4.90000
25%     0.750000   0.925000            10.77500
50%     8.000000   3.350000            30.95000
75%    35.250000  11.925000            67.15000
max    79.000000  30.100000           139.80000

In [22]:
df.sort_values(by = 'Diameter (in 1Kkm)', ascending = False)

Name  Satellite    AU  Diameter (in 1Kkm)
5  Jupiter         79   5.2               139.8
6   Saturn         60   9.5               116.5
7   Uranus         27  19.2                50.7
8  Neptune         14  30.1                49.2
3    Earth          1   1.0                12.7
2    Venus          0   0.7                12.1
4     Mars          2   1.5                 6.8
1  Mercury          0   0.4                 4.9

In [ ]:
# TO DO: re-sort the DataFrame by the number of satellites in descending order.
df.sort_values(None)

Before 2006, Pluto was classified as a planet of the solar system. Let's bring it back to our solar system, by adding Pluto to our DataFrame.

In [25]:
df.loc[9] = ["Pluto", 0, 39.5, 2.38]
df

Name  Satellite    AU  Diameter (in 1Kkm)
1  Mercury          0   0.4                4.90
2    Venus          0   0.7               12.10
3    Earth          1   1.0               12.70
4     Mars          2   1.5                6.80
5  Jupiter         79   5.2              139.80
6   Saturn         60   9.5              116.50
7   Uranus         27  19.2               50.70
8  Neptune         14  30.1               49.20
9    Pluto          0  39.5                2.38

Let's reclassify Pluto as a dwarf planet again.

In [ ]:
# To DO: Drop Pluto, you can do that with either df.drop(index=idx) or df.drop(df.index[idx]), where idx is Pluto's index
df.drop(None)

### 2.2 Read Data with Pandas
So far, we played with a small, example dataset. Pandas supports loading, reading, and writing data from/to various file format, including CSV, JSON and SQL, by converting it to a DataFrame. 
1. `pd.read_csv()` : Read CSV files
2. `pd.read_json()` : Read JSON files
3. `pd.read_sql_query()` : Read SQL files

Let's experiment with these:

In [67]:
# TODO: Try each option, and see what the difference is.
# Option 1
movie = pd.read_csv("./data/IMDB-Movie-Data.csv", index_col = "Title")

#Option 2
# movie = pd.read_csv("./data/IMDB-Movie-Data.csv")
print(type(movie))
movie

<class 'pandas.core.frame.DataFrame'>


Rank                     Genre  \
Title                                                     
Guardians of the Galaxy     1   Action,Adventure,Sci-Fi   
Prometheus                  2  Adventure,Mystery,Sci-Fi   
Split                       3           Horror,Thriller   
Sing                        4   Animation,Comedy,Family   
Suicide Squad               5  Action,Adventure,Fantasy   
...                       ...                       ...   
Secret in Their Eyes      996       Crime,Drama,Mystery   
Hostel: Part II           997                    Horror   
Step Up 2: The Streets    998       Drama,Music,Romance   
Search Party              999          Adventure,Comedy   
Nine Lives               1000     Comedy,Family,Fantasy   

                                                               Description  \
Title                                                                        
Guardians of the Galaxy  A group of intergalactic criminals are forced ...   
Prometheus               Following clues to the origin of mankind, a te...   
Split                    Three girls are kidnapped by a man with a diag...   
Sing                     In a city of humanoid animals, a hustling thea...   
Suicide Squad            A secret government agency recruits some of th...   
...                                                                    ...   
Secret in Their Eyes     A tight-knit team of rising investigators, alo...   
Hostel: Part II          Three American college students studying abroa...   
Step Up 2: The Streets   Romantic sparks occur between two dance studen...   
Search Party             A pair of friends embark on a mission to reuni...   
Nine Lives               A stuffy businessman finds himself trapped ins...   

                                     Director  \
Title                                           
Guardians of the Galaxy            James Gunn   
Prometheus                       Ridley Scott   
Split                      M. Night Shyamalan   
Sing                     Christophe Lourdelet   
Suicide Squad                      David Ayer   
...                                       ...   
Secret in Their Eyes                Billy Ray   
Hostel: Part II                      Eli Roth   
Step Up 2: The Streets             Jon M. Chu   
Search Party                   Scot Armstrong   
Nine Lives                   Barry Sonnenfeld   

                                                                    Actors  \
Title                                                                        
Guardians of the Galaxy  Chris Pratt, Vin Diesel, Bradley Cooper, Zoe S...   
Prometheus               Noomi Rapace, Logan Marshall-Green, Michael Fa...   
Split                    James McAvoy, Anya Taylor-Joy, Haley Lu Richar...   
Sing                     Matthew McConaughey,Reese Witherspoon, Seth Ma...   
Suicide Squad            Will Smith, Jared Leto, Margot Robbie, Viola D...   
...                                                                    ...   
Secret in Their Eyes     Chiwetel Ejiofor, Nicole Kidman, Julia Roberts...   
Hostel: Part II          Lauren German, Heather Matarazzo, Bijou Philli...   
Step Up 2: The Streets   Robert Hoffman, Briana Evigan, Cassie Ventura,...   
Search Party             Adam Pally, T.J. Miller, Thomas Middleditch,Sh...   
Nine Lives               Kevin Spacey, Jennifer Garner, Robbie Amell,Ch...   

                         Year  Runtime (Minutes)  Rating   Votes  \
Title                                                              
Guardians of the Galaxy  2014                121     8.1  757074   
Prometheus               2012                124     7.0  485820   
Split                    2016                117     7.3  157606   
Sing                     2016                108     7.2   60545   
Suicide Squad            2016                123     6.2  393727   
...                       ...                ...     ...     ...   
Secret in Their Eyes     2015                

In [ ]:
# To Do: Try the other two options
movie.iloc[2]
# movie.loc["Split"]
# movie.loc[2]    --> Is this going to work? if not, why not?

In [ ]:
# To Do: Sort the table by Ratings, in descending order
# Do you agree with the rankings? :)
movie.None

In [ ]:
# To Do: Sort the table by 'Revenue(Millions)', in ascending order and print the first 3 rows out
movie.None

In [28]:
# The value_counts() function is used to get a Series containing counts of unique values
movie['Genre'].value_counts().head()

Action,Adventure,Sci-Fi    50
Drama                      48
Comedy,Drama,Romance       35
Comedy                     32
Drama,Romance              31
Name: Genre, dtype: int64

In [29]:
# This is called a "Masking Operation"
# filter out movies that have runtime under 170 minutes and sort the result by rating in descending order.
movie[movie['Runtime (Minutes)'] >= 170].sort_values(by="Rating", ascending=False)

Rank                   Genre  \
Title                                                   
3 Idiots                  431            Comedy,Drama   
The Wolf of Wall Street    83  Biography,Comedy,Crime   
The Hateful Eight          89     Crime,Drama,Mystery   
La vie d'Adèle            312           Drama,Romance   
Grindhouse                829  Action,Horror,Thriller   
Cloud Atlas               268            Drama,Sci-Fi   
Inland Empire             966  Drama,Mystery,Thriller   

                                                               Description  \
Title                                                                        
3 Idiots                 Two friends are searching for their long lost ...   
The Wolf of Wall Street  Based on the true story of Jordan Belfort, fro...   
The Hateful Eight        In the dead of a Wyoming winter, a bounty hunt...   
La vie d'Adèle           Adèle's life is changed when she meets Emma, a...   
Grindhouse               Quentin Tarantino and Robert Rodriguez's homag...   
Cloud Atlas              An exploration of how the actions of individua...   
Inland Empire            As an actress starts to adopt the persona of h...   

                                    Director  \
Title                                          
3 Idiots                     Rajkumar Hirani   
The Wolf of Wall Street      Martin Scorsese   
The Hateful Eight          Quentin Tarantino   
La vie d'Adèle           Abdellatif Kechiche   
Grindhouse                  Robert Rodriguez   
Cloud Atlas                       Tom Tykwer   
Inland Empire                    David Lynch   

                                                                    Actors  \
Title                                                                        
3 Idiots                   Aamir Khan, Madhavan, Mona Singh, Sharman Joshi   
The Wolf of Wall Street  Leonardo DiCaprio, Jonah Hill, Margot Robbie,M...   
The Hateful Eight        Samuel L. Jackson, Kurt Russell, Jennifer Jaso...   
La vie d'Adèle           Léa Seydoux, Adèle Exarchopoulos, Salim Kechio...   
Grindhouse               Kurt Russell, Rose McGowan, Danny Trejo, Zoë Bell   
Cloud Atlas               Tom Hanks, Halle Berry, Hugh Grant, Hugo Weaving   
Inland Empire            Laura Dern, Jeremy Irons, Justin Theroux, Karo...   

                         Year  Runtime (Minutes)  Rating   Votes  \
Title                                                              
3 Idiots                 2009                170     8.4  238789   
The Wolf of Wall Street  2013                180     8.2  865134   
The Hateful Eight        2015                187     7.8  341170   
La vie d'Adèle           2013                180     7.8  103150   
Grindhouse               2007                191     7.6  160350   
Cloud Atlas              2012                172     7.5  298651   
Inland Empire            2006                180     7.0   44227   

                         Revenue (Millions)  Metascore  
Title                                                   
3 Idiots                               6.52       67.0  
The Wolf of Wall Street              116.87       75.0  
The Hateful Eight                     54.12       68.0  
La vie d'Adèle                         2.20       88.0  
Grindhouse                            25.03        NaN  
Cloud Atlas                           27.10       55.0  
Inland Empire                           NaN        NaN

In [ ]:
# To Do: By using a masking operation, Extract the movies whose 'Metascore' is bigger than 95, 
# and sort the result from the most recent to the least recent ones
None

In [ ]:
# To Do: Extract movies who are directed by one of UCL's alumni ---> Hint: Tenet, Inception
None

#### 2.2.1 Pandas Exercise

To Do: Extract the movie list that meets these requirements:
- 1. Released after 2010 (key = 'Year') (including the year 2010)
- 2. Runtime is shorter than 150 minutes (key = 'Runtime (Minutes)')
- 3. Rating is above 8.0 (key = 'Rating')  

Print out only the first 3 movies from the result.

In [ ]:
None

### 2.3 How to deal with Missing Data
To represent missing data, Pandas uses `np.nan` (this is the `np` from the NumPy tutorial!). Data scientists and machine learning engineers sometimes just remove missing data. However, it heavily depends on which data are missing, how big the missing data are and so on. You can fill the missing part with 0, with the mean value of the column, with the mean value of only the 10 closests values in the column or anything else that might seem appropriate. It is important for you to choose the way you are going to deal with missing data. Here are some methods to help you with that:
- `isnull()`: returns True or False, depending on the cell's null status. 
- `sum()`: This can be used as a trick when you count the number of Trues. Once the Dataframe is filtered through the `isnull()` function, the sum of all Trues in a column gives you how many fields have missing data in them.
- `dropna()`: deletes any row that contains at least one null value.
- `fillna(value)`: Fills missing values with the given values.

In [31]:
movie.isnull()

Rank  Genre  Description  Director  Actors   Year  \
Title                                                                         
Guardians of the Galaxy  False  False        False     False   False  False   
Prometheus               False  False        False     False   False  False   
Split                    False  False        False     False   False  False   
Sing                     False  False        False     False   False  False   
Suicide Squad            False  False        False     False   False  False   
...                        ...    ...          ...       ...     ...    ...   
Secret in Their Eyes     False  False        False     False   False  False   
Hostel: Part II          False  False        False     False   False  False   
Step Up 2: The Streets   False  False        False     False   False  False   
Search Party             False  False        False     False   False  False   
Nine Lives               False  False        False     False   False  False   

                         Runtime (Minutes)  Rating  Votes  Revenue (Millions)  \
Title                                                                           
Guardians of the Galaxy              False   False  False               False   
Prometheus                           False   False  False               False   
Split                                False   False  False               False   
Sing                                 False   False  False               False   
Suicide Squad                        False   False  False               False   
...                                    ...     ...    ...                 ...   
Secret in Their Eyes                 False   False  False                True   
Hostel: Part II                      False   False  False               False   
Step Up 2: The Streets               False   False  False               False   
Search Party                         False   False  False                True   
Nine Lives                           False   False  False               False   

                         Metascore  
Title                               
Guardians of the Galaxy      False  
Prometheus                   False  
Split                        False  
Sing                         False  
Suicide Squad                False  
...                            ...  
Secret in Their Eyes         False  
Hostel: Part II              False  
Step Up 2: The Streets       False  
Search Party                 False  
Nine Lives                   False  

[1000 rows x 11 columns]

In [32]:
movie.isnull().sum()

Rank                    0
Genre                   0
Description             0
Director                0
Actors                  0
Year                    0
Runtime (Minutes)       0
Rating                  0
Votes                   0
Revenue (Millions)    128
Metascore              64
dtype: int64

In [68]:
# We create a copy of the movie object, as we want to keep the original how we initially had it.
copy_of_movie = movie.copy()
# Take a look at "Take Me Home Tonight" and "Search Party"
copy_of_movie.fillna(value = 0)

Rank                     Genre  \
Title                                                     
Guardians of the Galaxy     1   Action,Adventure,Sci-Fi   
Prometheus                  2  Adventure,Mystery,Sci-Fi   
Split                       3           Horror,Thriller   
Sing                        4   Animation,Comedy,Family   
Suicide Squad               5  Action,Adventure,Fantasy   
...                       ...                       ...   
Secret in Their Eyes      996       Crime,Drama,Mystery   
Hostel: Part II           997                    Horror   
Step Up 2: The Streets    998       Drama,Music,Romance   
Search Party              999          Adventure,Comedy   
Nine Lives               1000     Comedy,Family,Fantasy   

                                                               Description  \
Title                                                                        
Guardians of the Galaxy  A group of intergalactic criminals are forced ...   
Prometheus               Following clues to the origin of mankind, a te...   
Split                    Three girls are kidnapped by a man with a diag...   
Sing                     In a city of humanoid animals, a hustling thea...   
Suicide Squad            A secret government agency recruits some of th...   
...                                                                    ...   
Secret in Their Eyes     A tight-knit team of rising investigators, alo...   
Hostel: Part II          Three American college students studying abroa...   
Step Up 2: The Streets   Romantic sparks occur between two dance studen...   
Search Party             A pair of friends embark on a mission to reuni...   
Nine Lives               A stuffy businessman finds himself trapped ins...   

                                     Director  \
Title                                           
Guardians of the Galaxy            James Gunn   
Prometheus                       Ridley Scott   
Split                      M. Night Shyamalan   
Sing                     Christophe Lourdelet   
Suicide Squad                      David Ayer   
...                                       ...   
Secret in Their Eyes                Billy Ray   
Hostel: Part II                      Eli Roth   
Step Up 2: The Streets             Jon M. Chu   
Search Party                   Scot Armstrong   
Nine Lives                   Barry Sonnenfeld   

                                                                    Actors  \
Title                                                                        
Guardians of the Galaxy  Chris Pratt, Vin Diesel, Bradley Cooper, Zoe S...   
Prometheus               Noomi Rapace, Logan Marshall-Green, Michael Fa...   
Split                    James McAvoy, Anya Taylor-Joy, Haley Lu Richar...   
Sing                     Matthew McConaughey,Reese Witherspoon, Seth Ma...   
Suicide Squad            Will Smith, Jared Leto, Margot Robbie, Viola D...   
...                                                                    ...   
Secret in Their Eyes     Chiwetel Ejiofor, Nicole Kidman, Julia Roberts...   
Hostel: Part II          Lauren German, Heather Matarazzo, Bijou Philli...   
Step Up 2: The Streets   Robert Hoffman, Briana Evigan, Cassie Ventura,...   
Search Party             Adam Pally, T.J. Miller, Thomas Middleditch,Sh...   
Nine Lives               Kevin Spacey, Jennifer Garner, Robbie Amell,Ch...   

                         Year  Runtime (Minutes)  Rating   Votes  \
Title                                                              
Guardians of the Galaxy  2014                121     8.1  757074   
Prometheus               2012                124     7.0  485820   
Split                    2016                117     7.3  157606   
Sing                     2016                108     7.2   60545   
Suicide Squad            2016                123     6.2  393727   
...                       ...                ...     ...     ...   
Secret in Their Eyes     2015                

In [69]:
# Important!!! This changes the copy_of_movie object itself!
copy_of_movie.dropna(inplace = True)

In [33]:
movie.shape

(1000, 11)

In [71]:
copy_of_movie.shape

(838, 11)

After dropping the rows that contain missing data, the shape of the data frame has changed, from (1000, 11) to (838, 11).

### 2.4 Merging Data
Those of you who know SQL might have felt that Pandas is quite similar to a query language.
What are the most common things that you do in most of the relational database query languages?  
Yes! (terminology alert!) An inner JOIN, an outer JOIN, a left JOIN, a right JOIN, a full JOIN, etc.
- `concat()` : Concatenation. Used to merge two or more Pandas object.
- `merge()` : Behaves very simlarly to SQL.

We'll create two random dataframes, named `df1` and `df2`.

In [ ]:
df1 = pd.DataFrame(np.random.randn(10, 2))
df1

In [ ]:
df2 = pd.DataFrame(np.random.randn(10, 3))
df2

In [ ]:
pd.concat([df1, df2])

In [ ]:
pd.concat([df1, df2], axis = 1)     # axis setting is very common in Pandas

In [37]:
demis = pd.DataFrame(
    {'Modules': ['Bioinformatics', 'Robotic Systems', 'Security', 'Compilers'], 'Demis' : [75, 97, 64, 81]}
)
demis

Modules  Demis
0   Bioinformatics     75
1  Robotic Systems     97
2         Security     64
3        Compilers     81

In [38]:
sedol = pd.DataFrame(
    {'Modules': ['Bioinformatics', 'Robotic Systems', 'Security', 'Compilers'], 'Sedol' : [63, 78, 84, 95]})
sedol

Modules  Sedol
0   Bioinformatics     63
1  Robotic Systems     78
2         Security     84
3        Compilers     95

In [39]:
pd.merge(demis, sedol, on = 'Modules')

Modules  Demis  Sedol
0   Bioinformatics     75     63
1  Robotic Systems     97     78
2         Security     64     84
3        Compilers     81     95

### 2.5 Exercise (Optional)

In [80]:
# Create the totally realistic popcorn dataset. The fact that the number of popcorn sold turns out
# to be the length of the name of the director times 100 000 is a total mystery to us too.
movie_popcorn = movie[["Director", "Description"]].copy()
movie_popcorn['Popcorn Sold'] = movie["Director"].str.len() * 1e5

In [81]:
movie_popcorn

Director  \
Title                                           
Guardians of the Galaxy            James Gunn   
Prometheus                       Ridley Scott   
Split                      M. Night Shyamalan   
Sing                     Christophe Lourdelet   
Suicide Squad                      David Ayer   
...                                       ...   
Secret in Their Eyes                Billy Ray   
Hostel: Part II                      Eli Roth   
Step Up 2: The Streets             Jon M. Chu   
Search Party                   Scot Armstrong   
Nine Lives                   Barry Sonnenfeld   

                                                               Description  \
Title                                                                        
Guardians of the Galaxy  A group of intergalactic criminals are forced ...   
Prometheus               Following clues to the origin of mankind, a te...   
Split                    Three girls are kidnapped by a man with a diag...   
Sing                     In a city of humanoid animals, a hustling thea...   
Suicide Squad            A secret government agency recruits some of th...   
...                                                                    ...   
Secret in Their Eyes     A tight-knit team of rising investigators, alo...   
Hostel: Part II          Three American college students studying abroa...   
Step Up 2: The Streets   Romantic sparks occur between two dance studen...   
Search Party             A pair of friends embark on a mission to reuni...   
Nine Lives               A stuffy businessman finds himself trapped ins...   

                         Popcorn Sold  
Title                                  
Guardians of the Galaxy     1000000.0  
Prometheus                  1200000.0  
Split                       1800000.0  
Sing                        2000000.0  
Suicide Squad               1000000.0  
...                               ...  
Secret in Their Eyes         900000.0  
Hostel: Part II              800000.0  
Step Up 2: The Streets      1000000.0  
Search Party                1400000.0  
Nine Lives                  1600000.0  

[1000 rows x 3 columns]

Above we created a fake dataset that tells us how much popcorn in total was sold for each movie. We would like you to find out what the average number of popcorn sold at movies that came out after 2014 is. **Hint**: you will need to join this new table with the existing movies table on the column `Description`. What happens to the size of the data frame if you try to join on the column `Director`? Can you guess why that is?

In [82]:
# To Do: Your code here

### What to do next?
The websites below are helpful for your further study of Pandas:
- [Pandas official website](https://pandas.pydata.org)
- [10 minutes to Pandas](https://pandas.pydata.org/pandas-docs/stable/getting_started/10min.html)
- [Data Wrangling with Pandas Cheat Sheet](https://pandas.pydata.org/Pandas_Cheat_Sheet.pdf)